app/routes.py

```diff
@@ -1,2 +1,289 @@
+
+    db.session.commit()
+
+    slack_bot(task.title)
+
+    response_body = {
+        "task": {
+            "id": task.task_id,
+            "title": task.title,
+            "description": task.description,
+            "is_complete": task.completed_at is not None
+        }
+    }
+    return jsonify(response_body), 200
+
+
+@tasks_bp.route("/<task_id>/mark_incomplete", methods=["PATCH"])
+def mark_incomplete(task_id):
+    task = Task.query.get_or_404(task_id)
+    task.completed_at = None
+
+    db.session.commit()
+
+    response_body = {
+        "task": {
+            "id": task.task_id,
+            "title": task.title,
+            "description": task.description,
+            "is_complete": task.completed_at is not None
+        }
+    }
+    return jsonify(response_body), 200
+
+
+@goals_bp.route("", methods=["POST"])
+def create_goal():
+    request_body = request.get_json()
+
+    if "title" not in request_body:
+        return jsonify({"details": "Invalid data"}), 400
+
+    new_goal = Goal(title=request_body["title"])
+
+    db.session.add(new_goal)
+    db.session.commit()
+
+    response_body = {
+        "goal": {
+            "id": new_goal.goal_id,
+            "title": new_goal.title
+        }
+    }
+    return jsonify(response_body), 201
+
+
+@goals_bp.route("", methods=["GET"])
+def read_goal():
+
+    goals = Goal.query.all()
+
+    goal_responses = []
+    for goal in goals:
+        goal_responses.append({
+            "id": goal.goal_id,
+            "title": goal.title
+        })
+
+    return jsonify(goal_responses), 200
+
+
+@goals_bp.route("/<goal_id>", methods=["GET"])
+def read_one_goal(goal_id):
+    goal = Goal.query.get_or_404(goal_id)
+
+    response_body = {
+        "goal": {
+            "id": goal.goal_id,
+            "title": goal.title
+        }
+    }
+    return jsonify(response_body), 200
+
+
+@goals_bp.route("/<goal_id>", methods=["PUT"])
+def update_one_goal(goal_id):
+    goal = Goal.query.get_or_404(goal_id)
+
+    request_body = request.get_json()
+
+    goal.title = request_body["title"]
+
+    db.session.commit()
+
+    response_body = {
+        "goal": {
+            "id": goal.goal_id,
+            "title": goal.title,
+        }
+    }
+    return jsonify(response_body), 200
+
+
+@goals_bp.route("/<goal_id>", methods=["DELETE"])
+def delete_one_goal(goal_id):
+    goal = Goal.query.get_or_404(goal_id)
+
+    db.session.delete(goal)
+    db.session.commit()
+
+    return jsonify({
+        "details": f"Goal {goal.goal_id} \"{goal.title}\" successfully deleted"
+    }), 200
+
+
+@goals_bp.route("/<goal_id>/tasks", methods=["POST"])
+def post_tasks_to_goal(goal_id):
+    request_body = request.get_json()
+
+    if "task_ids" not in request_body:
+        return jsonify({"details": "Invalid data"}), 400
+
+    task_ids = request_body["task_ids"]
+    for task_id in task_ids:
+        task = Task.query.get_or_404(task_id)
+        task.goal_id = goal_id
```

Comment:
👍 

This is absolutely a valid way to associate a task with a goal. But we can also think about using either end of the goal → tasks relationship we created in the models to help.

One approach we could do if the `tasks` relationship had a `backref="goal"`, which is essentially equivalent to what is written here would be
```python
            task.goal = goal  # use actual model instances rather than the raw ids. `goal` is added to the task by the db.relationship
```

Or to go from the opposite side of the relationship, we could write
```python
            goal.tasks.append(task)  # append the looked up task to the tasks collection of the goal. tasks is added to the goal by the db.relationship
```

In either situation, we should also think about whether to interpret POSTing to this endpoint as setting the exact list of tasks that should be a part of the goal, or whether we are adding these tasks to the goal (in addition to anything that was already there). Neither the project description nor the test `test_post_task_ids_to_goal_already_with_goals` are entirely clear on the desired behavior. Which perspective does your implementation take? How could we implement the other approach?